File: CoreVia/CoreVia/Features/AICalorie/Services/FoodDatabaseService.swift
//
//  FoodDatabaseService.swift
//  CoreVia
//
//  On-device qida verilənlər bazası servisi
//  USDA + Azerbaycan yeməkləri — offline nutrition lookup
//  Backend food_database.py ilə eyni matching logic
//

import Foundation

// MARK: - Models

struct FoodNutritionInfo {
    let foodName: String
    let caloriesPer100g: Double
    let proteinPer100g: Double
    let carbsPer100g: Double
    let fatPer100g: Double
    let portionGrams: Double
    let portionDesc: String
}

struct PortionNutrition {
    let foodName: String
    let calories: Int
    let protein: Double
    let carbs: Double
    let fat: Double
    let portionGrams: Double
    let portionDesc: String
    let confidence: Double
    let matched: Bool
}

// MARK: - Service

class FoodDatabaseService {
    static let shared = FoodDatabaseService()

    private var foods: [String: FoodNutritionInfo] = [:]
    private var foodNames: [String] = []

    private init() {
        loadDatabase()
    }

    // MARK: - Load

    private func loadDatabase() {
        guard let url = Bundle.main.url(forResource: "food_database", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: [String: Any]]
        else {
            print("⚠️ food_database.json yuklenmedi")
            return
        }

        for (name, info) in json {
            // Həm orijinal, həm lowercase key saxla — lookup asanlaşır
            let lowerName = name.lowercased()
            let displayName = name.replacingOccurrences(of: "_", with: " ")
                .split(separator: " ").map { $0.capitalized }.joined(separator: " ")
            let entry = FoodNutritionInfo(
                foodName: displayName,
                caloriesPer100g: (info["calories"] as? NSNumber)?.doubleValue ?? 0,
                proteinPer100g: (info["protein"] as? NSNumber)?.doubleValue ?? 0,
                carbsPer100g: (info["carbs"] as? NSNumber)?.doubleValue ?? 0,
                fatPer100g: (info["fat"] as? NSNumber)?.doubleValue ?? 0,
                portionGrams: (info["portion_g"] as? NSNumber)?.doubleValue ?? 200,
                portionDesc: info["portion_desc"] as? String ?? "1 portion"
            )
            // Orijinal key
            foods[name] = entry
            // Lowercase key (əgər fərqlidirsə)
            if lowerName != name { foods[lowerName] = entry }
            // Underscore variant
            let underscoreKey = lowerName.replacingOccurrences(of: " ", with: "_")
            if underscoreKey != lowerName { foods[underscoreKey] = entry }
            // Space variant
            let spaceKey = lowerName.replacingOccurrences(of: "_", with: " ")
            if spaceKey != lowerName { foods[spaceKey] = entry }
        }
        foodNames = Array(foods.keys)
        print("✅ FoodDatabase: \(foods.count) qida yuklendi")
    }

    // MARK: - Lookup

    /// Qida adına görə beslenme dəyərlərini qaytarır (porsiya üçün hesablanmış)
    func getNutrition(for foodName: String) -> PortionNutrition {
        let query = foodName.lowercased().trimmingCharacters(in: .whitespaces)
        // Underscore ↔ space normalization (model "bell_pepper" qaytara bilər, DB "Bell Pepper" saxlayır)
        let queryUnderscore = query.replacingOccurrences(of: " ", with: "_")
        let querySpace = query.replacingOccurrences(of: "_", with: " ")

        // 1. Exact match (3 variant: original, underscore, space)
        for variant in [query, queryUnderscore, querySpace] {
            if let food = foods[variant] {
                return calculatePortion(food)
            }
        }

        // 2. Partial match (hər iki format yoxla)
        for name in foodNames {
            let nameLower = name.lowercased()
            if nameLower.contains(query) || query.contains(nameLower)
                || nameLower.contains(querySpace) || querySpace.contains(nameLower)
                || nameLower.contains(queryUnderscore) || queryUnderscore.contains(nameLower) {
                if let food = foods[name] {
                    return calculatePortion(food)
                }
            }
        }

        // 3. Fuzzy match (LCS similarity > 0.55)
        var bestMatch: String?
        var bestScore: Double = 0

        for name in foodNames {
            let score = stringSimilarity(query, name)
            if score > bestScore {
                bestScore = score
                bestMatch = name
            }
        }

        if let match = bestMatch, bestScore > 0.55, let food = foods[match] {
            return calculatePortion(food)
        }

        // 4. Default fallback
        return PortionNutrition(
            foodName: foodName.capitalized,
            calories: 200,
            protein: 10.0,
            carbs: 25.0,
            fat: 8.0,
            portionGrams: 200,
            portionDesc: "1 portion (~200g)",
            confidence: 0.3,
            matched: false
        )
    }

    // MARK: - Helpers

    private func calculatePortion(_ food: FoodNutritionInfo) -> PortionNutrition {
        let multiplier = food.portionGrams / 100.0
        return PortionNutrition(
            foodName: food.foodName.capitalized,
            calories: Int((food.caloriesPer100g * multiplier).rounded()),
            protein: (food.proteinPer100g * multiplier * 10).rounded() / 10,
            carbs: (food.carbsPer100g * multiplier * 10).rounded() / 10,
            fat: (food.fatPer100g * multiplier * 10).rounded() / 10,
            portionGrams: food.portionGrams,
            portionDesc: food.portionDesc,
            confidence: 0.9,
            matched: true
        )
    }

    /// SequenceMatcher ekvivalenti — Longest Common Subsequence ratio
    private func stringSimilarity(_ a: String, _ b: String) -> Double {
        let aChars = Array(a)
        let bChars = Array(b)
        let aLen = aChars.count
        let bLen = bChars.count
        guard aLen > 0, bLen > 0 else { return 0 }

        // LCS dynamic programming
        var dp = Array(repeating: Array(repeating: 0, count: bLen + 1), count: aLen + 1)
        for i in 1...aLen {
            for j in 1...bLen {
                if aChars[i - 1] == bChars[j - 1] {
                    dp[i][j] = dp[i - 1][j - 1] + 1
                } else {
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
                }
            }
        }

        return Double(2 * dp[aLen][bLen]) / Double(aLen + bLen)
    }
}

File: CoreVia-Android/app/src/main/java/life/corevia/app/data/local/FoodDatabaseService.kt
package life.corevia.app.data.local

import android.content.Context
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.double
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.max
import kotlin.math.roundToInt

/**
 * On-device qida verilənlər bazası servisi
 * USDA + Azərbaycan yeməkləri — offline nutrition lookup
 * iOS FoodDatabaseService.swift ilə eyni matching logic
 */

data class FoodNutritionInfo(
    val foodName: String,
    val caloriesPer100g: Double,
    val proteinPer100g: Double,
    val carbsPer100g: Double,
    val fatPer100g: Double,
    val portionGrams: Double,
    val portionDesc: String
)

data class PortionNutrition(
    val foodName: String,
    val calories: Int,
    val protein: Double,
    val carbs: Double,
    val fat: Double,
    val portionGrams: Double,
    val portionDesc: String,
    val confidence: Double,
    val matched: Boolean
)

@Singleton
class FoodDatabaseService @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private val foods = mutableMapOf<String, FoodNutritionInfo>()
    private val foodNames = mutableListOf<String>()
    private var isLoaded = false

    init {
        loadDatabase()
    }

    // MARK: - Load

    private fun loadDatabase() {
        try {
            val jsonString = context.assets.open("food_database.json")
                .bufferedReader()
                .use { it.readText() }

            val json = Json { ignoreUnknownKeys = true }
            val jsonObject = json.decodeFromString<JsonObject>(jsonString)

            for ((name, value) in jsonObject) {
                val info = value.jsonObject
                val lowerName = name.lowercase()
                val displayName = name.replace("_", " ")
                    .split(" ")
                    .joinToString(" ") { it.replaceFirstChar { c -> c.uppercase() } }

                val entry = FoodNutritionInfo(
                    foodName = displayName,
                    caloriesPer100g = info["calories"]?.jsonPrimitive?.double ?: 0.0,
                    proteinPer100g = info["protein"]?.jsonPrimitive?.double ?: 0.0,
                    carbsPer100g = info["carbs"]?.jsonPrimitive?.double ?: 0.0,
                    fatPer100g = info["fat"]?.jsonPrimitive?.double ?: 0.0,
                    portionGrams = info["portion_g"]?.jsonPrimitive?.double ?: 200.0,
                    portionDesc = info["portion_desc"]?.jsonPrimitive?.content ?: "1 portion"
                )

                // Original key
                foods[name] = entry
                // Lowercase key
                if (lowerName != name) foods[lowerName] = entry
                // Underscore variant
                val underscoreKey = lowerName.replace(" ", "_")
                if (underscoreKey != lowerName) foods[underscoreKey] = entry
                // Space variant
                val spaceKey = lowerName.replace("_", " ")
                if (spaceKey != lowerName) foods[spaceKey] = entry
            }

            foodNames.addAll(foods.keys)
            isLoaded = true
            Timber.d("✅ FoodDatabase: ${foods.size} qida yükləndi")
        } catch (e: Exception) {
            Timber.e("⚠️ food_database.json yüklənmədi: ${e.message}")
        }
    }

    // MARK: - Lookup

    /**
     * Qida adına görə besləmə dəyərlərini qaytarır (porsiya üçün hesablanmış)
     * iOS FoodDatabaseService.getNutrition() ilə eyni logic
     */
    fun getNutrition(foodName: String): PortionNutrition {
        val query = foodName.lowercase().trim()
        val queryUnderscore = query.replace(" ", "_")
        val querySpace = query.replace("_", " ")

        // 1. Exact match (3 variant: original, underscore, space)
        for (variant in listOf(query, queryUnderscore, querySpace)) {
            foods[variant]?.let { return calculatePortion(it) }
        }

        // 2. Partial match
        for (name in foodNames) {
            val nameLower = name.lowercase()
            if (nameLower.contains(query) || query.contains(nameLower)
                || nameLower.contains(querySpace) || querySpace.contains(nameLower)
                || nameLower.contains(queryUnderscore) || queryUnderscore.contains(nameLower)
            ) {
                foods[name]?.let { return calculatePortion(it) }
            }
        }

        // 3. Fuzzy match (LCS similarity > 0.55)
        var bestMatch: String? = null
        var bestScore = 0.0

        for (name in foodNames) {
            val score = stringSimilarity(query, name)
            if (score > bestScore) {
                bestScore = score
                bestMatch = name
            }
        }

        if (bestMatch != null && bestScore > 0.55) {
            foods[bestMatch]?.let { return calculatePortion(it) }
        }

        // 4. Default fallback
        return PortionNutrition(
            foodName = foodName.replaceFirstChar { it.uppercase() },
            calories = 200,
            protein = 10.0,
            carbs = 25.0,
            fat = 8.0,
            portionGrams = 200.0,
            portionDesc = "1 portion (~200g)",
            confidence = 0.3,
            matched = false
        )
    }

    // MARK: - Helpers

    private fun calculatePortion(food: FoodNutritionInfo): PortionNutrition {
        val multiplier = food.portionGrams / 100.0
        return PortionNutrition(
            foodName = food.foodName,
            calories = (food.caloriesPer100g * multiplier).roundToInt(),
            protein = (food.proteinPer100g * multiplier * 10).roundToInt() / 10.0,
            carbs = (food.carbsPer100g * multiplier * 10).roundToInt() / 10.0,
            fat = (food.fatPer100g * multiplier * 10).roundToInt() / 10.0,
            portionGrams = food.portionGrams,
            portionDesc = food.portionDesc,
            confidence = 0.9,
            matched = true
        )
    }

    /**
     * SequenceMatcher ekvivalenti — Longest Common Subsequence ratio
     * iOS-dakı stringSimilarity() ilə eyni logic
     */
    private fun stringSimilarity(a: String, b: String): Double {
        val aLen = a.length
        val bLen = b.length
        if (aLen == 0 || bLen == 0) return 0.0

        // LCS dynamic programming
        val dp = Array(aLen + 1) { IntArray(bLen + 1) }
        for (i in 1..aLen) {
            for (j in 1..bLen) {
                dp[i][j] = if (a[i - 1] == b[j - 1]) {
                    dp[i - 1][j - 1] + 1
                } else {
                    max(dp[i - 1][j], dp[i][j - 1])
                }
            }
        }

        return (2.0 * dp[aLen][bLen]) / (aLen + bLen)
    }

    fun isReady(): Boolean = isLoaded
}
